lib/src/model/study/study_controller.dart

```diff
@@ -0,0 +1,628 @@
+import 'dart:async';
+
+import 'package:chessground/chessground.dart';
+import 'package:collection/collection.dart';
+import 'package:dartchess/dartchess.dart';
+import 'package:fast_immutable_collections/fast_immutable_collections.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:lichess_mobile/src/model/analysis/analysis_preferences.dart';
+import 'package:lichess_mobile/src/model/common/chess.dart';
+import 'package:lichess_mobile/src/model/common/eval.dart';
+import 'package:lichess_mobile/src/model/common/id.dart';
+import 'package:lichess_mobile/src/model/common/node.dart';
+import 'package:lichess_mobile/src/model/common/service/move_feedback.dart';
+import 'package:lichess_mobile/src/model/common/service/sound_service.dart';
+import 'package:lichess_mobile/src/model/common/uci.dart';
+import 'package:lichess_mobile/src/model/engine/evaluation_service.dart';
+import 'package:lichess_mobile/src/model/engine/work.dart';
+import 'package:lichess_mobile/src/model/study/study.dart';
+import 'package:lichess_mobile/src/model/study/study_repository.dart';
+import 'package:lichess_mobile/src/utils/rate_limit.dart';
+import 'package:lichess_mobile/src/view/engine/engine_gauge.dart';
+import 'package:lichess_mobile/src/widgets/pgn.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'study_controller.freezed.dart';
+part 'study_controller.g.dart';
+
+@riverpod
+class StudyController extends _$StudyController implements PgnTreeNotifier {
+  late Root _root;
+
+  final _engineEvalDebounce = Debouncer(const Duration(milliseconds: 150));
+
+  Timer? _startEngineEvalTimer;
+
+  Future<void> nextChapter() async {
+    if (state.hasValue) {
+      final chapters = state.requireValue.study.chapters;
+      final currentChapterIndex = chapters.indexWhere(
+        (chapter) => chapter.id == state.requireValue.study.chapter.id,
+      );
+      goToChapter(chapters[currentChapterIndex + 1].id);
+    }
+  }
+
+  Future<void> goToChapter(StudyChapterId chapterId) async {
+    state = const AsyncValue.loading();
+    state = AsyncValue.data(
+      await _fetchChapter(
+        state.requireValue.study.id,
+        chapterId: chapterId,
+      ),
+    );
+  }
+
+  Future<StudyState> _fetchChapter(
+    StudyId id, {
+    StudyChapterId? chapterId,
+  }) async {
+    final (study, pgn) = await ref
+        .read(studyRepositoryProvider)
+        .getStudy(id: id, chapterId: chapterId);
+
+    final game = PgnGame.parsePgn(pgn);
+
+    final rootComments = IList(game.comments.map((c) => PgnComment.fromPgn(c)));
+
+    final variant = study.chapter.setup.variant;
+    final orientation = study.chapter.setup.orientation;
+
+    try {
+      _root = Root.fromPgnGame(game);
+    } on PositionSetupException {
+      return StudyState(
```

Comment:
How does the web client handle these situations? Should we display a snackbar warning?

Reply:
After seeing the comment in `StudyState`, I realise this the way to handle illegal positions that can normally happen in study chapters. I think it deserves a comment to explain it to future readers.